ibis/formats/pyarrow.py

```diff
@@ -341,3 +346,46 @@ def to_polars(self, schema: Schema) -> pl.DataFrame:
 
         df = pl.from_arrow(self.obj)
         return PolarsData.convert_table(df, schema)
+
+
+class PyArrowDatasetProxy(TableProxy):
+    __slots__ = ("obj", "__dict__")
+    obj: V
+
+    def __init__(self, obj: V):
+        self.obj = obj
+
+    def __len__(self):
+        return self.obj.count_rows()
+
+    # pyarrow datasets are hashable, so we override the hash from TableProxy
+    def __hash__(self):
+        return hash(self.obj)
+
+    @cached_property
+    def _cache(self):
```

Comment:
If we keep around the `to_frame`/`to_pyarrow` codepaths, I'm a bit wary of caching here. For non-local backends the data will be used to create a temp table (meaning we don't need it in local memory anymore). I'd opt to remove the caching for now.